controller/src/main/java/io/pravega/controller/store/stream/ZKStreamMetadataStore.java

```diff
@@ -70,6 +97,82 @@ ZKStream newStream(final String scope, final String name) {
         return new ZKStream(scope, name, storeHelper);
     }
 
+    @Override
+    CompletableFuture<BigLong> getNextCounter() {
+        CompletableFuture<BigLong> future;
+        synchronized (lock) {
+            BigLong next = counter.incrementAndGet();
+            if (next.compareTo(limit.get()) > 0) {
```

Comment:
You are making this check here, and then again inside `refreshRangeIfNeeded`. Can we move most of the contents of this block inside that method?

Reply:
this block is simply checking if we need to refresh the range if we exceed the limit. if we are within limit we dont even call `refreshrangeifneeded`.
In `refreshRangeIfNeeded` we check to make sure only one refresh is ongoing. 
So if we have exceeded the limit, then there is definitely one refresh happening in the background. Once refresh completes, we will have the ranges updated.